src/backend/gl/src/command.rs

```diff
@@ -164,8 +153,12 @@ impl From<c::Limits> for Limits {
 /// `display_fb` field.
 #[derive(Clone)]
 pub struct RawCommandBuffer {
-    pub buf: Vec<Command>,
-    pub data: DataBuffer,
+    pub(crate) memory: Arc<RefCell<pool::OwnedBuffer>>,
```

Comment:
Unfortunately we need some way of sharing the underlying command buffer memory with the pool to allow resetting it (from the pool side), therefore an `Arc` wrapper. The common pattern is `Arc<Mutex<_>>` indeed but I wanted to avoid the locking overhead, which is not needed if our API is used correctly.

I'm may be tunnel visioning too much at this specific solution to see some easier approach.

Reply:
https://github.com/gfx-rs/gfx/pull/1461/commits/a8438c18ae4c0cdf70aa4db569d63038eadf0a4c
contains the earlier `*mut Vec<_>` implementation, but it's a bit more spooky as mentioned in the PR text.
In theory it's conformant to vulkan's error handling philosophy imo, but I went away from it as I'm unsure if we want to be that lenient for our GL backend.
> Except as documented elsewhere in the Specification, the behavior of the core layer to an application using the API incorrectly is undefined, and may include program termination. However, implementations must ensure that incorrect usage by an application does not affect the integrity of the operating system, the Vulkan implementation, or other Vulkan client applications in the system, and does not allow one application to access data belonging to another application.

Reply:
If you don't need `Send`/`Sync` why not use `Rc` ? Otherwise I don't think you will get `Send`/`Sync` with `Arc` either since `RefCell` is not `Sync`.

Reply:
`RawCommandBuffer` needs to `Clone` + `Send` but it would be also possible.
It's mainly a tradeoff about how unsafe we want to make the backend in the `undefined behavior` area.

An example scenario where `Arc` is better would be to prevent the memory being dropped (calling `free`) while a clone of a buffer is currently submitted to a queue. Due to `Send` it's possible that the memory gets dropped earlier as the refcount in the `Rc` _could_ be wrong..

Reply:
What do you mean it would be also possible ? Neither `Arc<RefCell<_>>` nor `Rc<RefCell<_>>` are going to be `Send` as far as I know. Otherwise you would end up with two references on the `RefCell` from different threads, which requires it to be `Sync`.

If sharing ownership is really necessary, I would instead use `Arc<UnsafeCell>` + an atomic to keep track of ownership. This is essentially what I did for the previous memory mapping tracking system where I needed a `Mutex`-like without lock.

Reply:
@Bastacyclop sounds reasonable. I wonder how the code compiled then?..
Speaking of which, we need to disable some crates in the CI scripts in order to start relying on it again for this branch.

Reply:
@Bastacyclop Manually marked it as `Send` as it's not by default as you stated correctly. `UnsafeCell` + atomic sounds actually interesting and I'll give it a try, thanks!